packages/sanity/src/core/studio/router/types.ts

```diff
@@ -11,3 +13,17 @@ export interface RouterIntentEvent {
 }
 
 export type RouterEvent = RouterStateEvent | RouterIntentEvent
+
+/**
+ * A subset of the History API is used, and explicitly declared so it's possible to write a custom
+ * history implementation that can be used to integrate the router in a variety of parent routers.
+ * @internal
+ */
+export type RequiredHistory = Pick<History, 'listen' | 'location' | 'push' | 'replace'>
```

Comment:
It's now a bit easier to do deep integrations with the `unstable_router` prop.

Passing any of `createBrowserHistory`, `createHashHistory` and `createMemoryHistory` directly to `unstable_history` still works the same:
```tsx
import {createMemoryHistory} from 'history'
import {Studio} from 'sanity'
import sanityConfig from '../../sanity.config'

const history = createMemoryHistory({initialEntries: [{pathname: '/test'}]})

function Main() {
  return (
    <Studio
      config={sanityConfig}
      unstable_history={history}
    />
  )
}
```

What's new is when you wish to make a deep integration to something like, say, NextJS:
```tsx
import { createMemoryHistory } from 'history'
import type { GetServerSideProps, InferGetServerSidePropsType } from 'next'
import { useRouter } from 'next/router'
import { NextStudio } from 'next-sanity/studio'
import { useEffect, useMemo, useRef, useState } from 'react'
import type { StudioProviderProps } from 'sanity'
import config from 'sanity.config'

export const getServerSideProps: GetServerSideProps = async ({
  resolvedUrl,
}) => ({
  props: { resolvedUrl },
})

export default function StudioPage(
  props: InferGetServerSidePropsType<typeof getServerSideProps>
) {
  const history = useNextHistory(props.resolvedUrl)
  return <NextStudio config={config} unstable_history={history} />
}

type RouterHistory = StudioProviderProps['unstable_history']
function useNextHistory(resolvedUrl: string): RouterHistory {
  const router = useRouter()
  const [initialEntries] = useState(() => [resolvedUrl])
  const memoryHistory = useMemo(
    () => createMemoryHistory({ initialEntries }),
    [initialEntries]
  )
  const routerRef = useRef(router)
  useEffect(() => {
    routerRef.current = router
    if (memoryHistory.location.pathname !== router.asPath) {
      memoryHistory.push(router.asPath)
    }
  }, [router, memoryHistory])

  return useMemo(() => {
    return {
      listen: (listener) => memoryHistory.listen(listener),
      get location() {
        return memoryHistory.location
      },
      push: (to): void => {
        routerRef.current.push(
          typeof to === 'string' ? decodeURIComponent(to) : to,
          undefined,
          { shallow: true }
        )
      },
      replace: (to): void => {
        routerRef.current.replace(
          typeof to === 'string' ? decodeURIComponent(to) : to,
          undefined,
          { shallow: true }
        )
      },
    } satisfies RouterHistory
  }, [memoryHistory])
}
```

The `RequiredHistory` typing makes it clear what `history` APIs we're actually using. So you won't waste your time implementing things like `history.createHref` and `history.block`.
With the NextJS sample above it's able to sync the Studio state with the NextJS router. Compared to previous approaches this means that using features like the forward/backwards browser nav buttons always works, instead of only sometimes.
And the `getServerSideProps` step ensures that `createMemoryHistory({ initialEntries })` looks like `createMemoryHistory({ initialEntries: ['/studio/desk'] })` if the current URL is `/studio/desk`. Skipping a few loading steps and fast tracking startup.